File: LeetCode/C/143-linked-list-重排链表.cpp
/*
 * @lc app=leetcode.cn id=143 lang=cpp
 *
 * [143] 重排链表
 */
#include "main.h"

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
  public:
    void reorderList(ListNode *head) {
        if (head == nullptr || head->next == nullptr)
            return;
        ListNode *right, *p, *t;
        // split
        right = head, p = head;
        while (p && p->next) {
            right = right->next;
            p = p->next->next;
        }
        p = right;
        right = right->next;
        p->next = nullptr;
        // reverse
        p = nullptr;
        while (right) {
            t = right->next;
            right->next = p;
            p = right;
            right = t;
        }
        right = p;
        // insert
        p = head;
        while (p && right) {
            t = right;
            right = right->next;
            t->next = p->next;
            p->next = t;
            p = t->next;
        }
    }
};
// @lc code=end

File: LeetCode/Go/143-linked-list-重排链表.go
/*
 * @lc app=leetcode.cn id=143 lang=golang
 *
 * [143] 重排链表
 */
package leetcode

// @lc code=start
func reorderList(head *ListNode) {
	if head == nil || head.Next == nil {
		return
	}
	// split
	var right, p = head, head
	for p != nil && p.Next != nil {
		right = right.Next
		p = p.Next.Next
	}
	p = right
	right = right.Next
	p.Next = nil
	// reverse
	p = nil
	for right != nil {
		var t = right.Next
		right.Next = p
		p = right
		right = t
	}
	right = p
	// insert
	p = head
	for p != nil && right != nil {
		var t = right
		right = right.Next
		t.Next = p.Next
		p.Next = t
		p = t.Next
	}
}

// @lc code=end
